storage/src/main/java/org/apache/kafka/server/log/internals/ProducerStateManager.java

```diff
@@ -0,0 +1,671 @@
+        for (SnapshotFile snapshot : snapshots.values()) {
+            if (snapshot.offset > logEndOffset || snapshot.offset <= logStartOffset) {
+                removeAndDeleteSnapshot(snapshot.offset);
+            }
+        }
+
+        if (logEndOffset != mapEndOffset()) {
+            producers.clear();
+            ongoingTxns.clear();
+            updateOldestTxnTimestamp();
+
+            // since we assume that the offset is less than or equal to the high watermark, it is
+            // safe to clear the unreplicated transactions
+            unreplicatedTxns.clear();
+            loadFromSnapshot(logStartOffset, currentTimeMs);
+        } else {
+            onLogStartOffsetIncremented(logStartOffset);
+        }
+    }
+
+    public ProducerAppendInfo prepareUpdate(long producerId, AppendOrigin origin) {
+        ProducerStateEntry currentEntry = lastEntry(producerId).orElse(ProducerStateEntry.empty(producerId));
+        return new ProducerAppendInfo(topicPartition, producerId, currentEntry, origin);
+    }
+
+
+    /**
+     * Update the mapping with the given append information
+     */
+    public void update(ProducerAppendInfo appendInfo) {
+        if (appendInfo.producerId() == RecordBatch.NO_PRODUCER_ID)
+            throw new IllegalArgumentException("Invalid producer id " + appendInfo.producerId() + " passed to update " + "for partition $topicPartition");
+
+        log.trace("Updated producer " + appendInfo.producerId() + " state to" + appendInfo);
+        ProducerStateEntry updatedEntry = appendInfo.toEntry();
+        ProducerStateEntry producerStateEntry = producers.get(appendInfo.producerId());
+        if (producerStateEntry != null) {
+            producerStateEntry.update(updatedEntry);
+        } else {
+            producers.put(appendInfo.producerId(), updatedEntry);
+        }
+
+        appendInfo.startedTransactions().forEach(txn -> ongoingTxns.put(txn.firstOffset.messageOffset, txn));
+
+        updateOldestTxnTimestamp();
+    }
+
+
+    private void updateOldestTxnTimestamp() {
+        Map.Entry<Long, TxnMetadata> firstEntry = ongoingTxns.firstEntry();
+        if (firstEntry == null) {
+            oldestTxnLastTimestamp = -1;
+        } else {
+            TxnMetadata oldestTxnMetadata = firstEntry.getValue();
+            ProducerStateEntry entry = producers.get(oldestTxnMetadata.producerId);
+            oldestTxnLastTimestamp = entry != null ? entry.lastTimestamp() : -1L;
+        }
+    }
+
+    public void updateMapEndOffset(long lastOffset) {
+        lastMapOffset = lastOffset;
+    }
+
+    /**
+     * Get the last written entry for the given producer id.
+     */
+    public Optional<ProducerStateEntry> lastEntry(long producerId) {
+        return Optional.ofNullable(producers.get(producerId));
+    }
+
+    /**
+     * Take a snapshot at the current end offset if one does not already exist.
+     */
+    public void takeSnapshot() throws IOException {
```

Comment:
We are trying to avoid making changes that could result in subtle breakage unless there is a strong reason to make the change and we have a high degree of confidence that it's safe to do so. With that in mind, what's the reason for making the change you are suggesting?

Reply:
Thanks for the follow-up. I missed to mention the question above is beyond the scope of this PR. I am trying to understand the strategy adopted in case of I/O failures in the data path. Some I/O failures directly results in putting the log directory written on offline. Some others are propagated up the stack. Is there a design pattern applying here that we have to follow? Do all I/O failures warrant the log directory to be made offline?

Reply:
Apologies - I just found an answer to this question [here](https://github.com/apache/kafka/blob/trunk/clients/src/main/java/org/apache/kafka/common/errors/KafkaStorageException.java#L23-L28).

Quoting from `KafkaStorageException`'s Javadoc:

> 1) If the server has not finished loading logs, IOException does not need to be converted to KafkaStorageException
> 2) After the server has finished loading logs, IOException should be caught and trigger LogDirFailureChannel.maybeAddOfflineLogDir(). Then the IOException should either be swallowed and logged, or be converted and re-thrown as KafkaStorageException
> 3) It is preferred for IOException to be caught in Log rather than in ReplicaManager or LogSegment.